python/psautohint/otfFont.py

```diff
@@ -1000,3 +1099,123 @@ def getfdInfo(self, allow_no_blues, noFlex, vCounterGlyphs, hCounterGlyphs,
+
+    def fix_glyph_hints(self, glyph_name, mm_hint_info, is_reference_font):
+        # 1. Delete any bad hints.
+        # 2. If reference font, recalculate the hint mask byte strings
+        # 3. Replace hint masks.
+        # 3. Fix cntr masks.
+        try:
+            t2CharString = self.charStrings[glyph_name]
+        except KeyError:
+            return  # Happens with sparse source fonts - just skip the glyph.
+        subrs = getattr(t2CharString.private, "Subrs", [])
+        decompiler = FixHintWidthDecompiler(subrs, t2CharString.globalSubrs,
+                                            t2CharString.private)
+        decompiler.execute(t2CharString)
+        program = t2CharString.program
+        num_hhint_pairs = len(decompiler.h_hint_args) // 2
+
+        # 1. Build list of good [vh]hints.
+        bad_hint_idxs = list(mm_hint_info.bad_hint_idxs)
+        bad_hint_idxs.sort()
+        for idx in reversed(bad_hint_idxs):
+            if idx < num_hhint_pairs:
+                hint_args = decompiler.h_hint_args
+                bottom_idx = idx * 2
+            else:
+                hint_args = decompiler.v_hint_args
+                bottom_idx = (idx-num_hhint_pairs) * 2
+            delta = hint_args[bottom_idx] + hint_args[bottom_idx + 1]
+            del hint_args[bottom_idx:bottom_idx+2]
+            if len(hint_args) > bottom_idx:
+                hint_args[bottom_idx] += delta
+
+        # delete old hints from program
+        if mm_hint_info.cntr_masks:
+            last_hint_idx = program.index('cntrmask')
+        elif mm_hint_info.hint_masks:
+            last_hint_idx = program.index('hintmask')
+        else:
+            for op in ['vstem', 'hstem']:
+                try:
+                    last_hint_idx = program.index(op)
+                    break
+                except IndexError:
+                    last_hint_idx = None
+        if last_hint_idx is not None:
+            del program[:last_hint_idx]
+
+        # If there were decompiler.v_hint_args, but they have now all been
+        # deleted, the first token will still be 'vstem[hm]'. Delete it.
+        if ((not decompiler.v_hint_args)
+                and program[0].startswith('vstem')):
+            del program[0]
+
+        # Add width and updated hints back.
+        if (decompiler.has_explicit_width):
+            hint_program = [decompiler.width_arg]
+        else:
+            hint_program = []
+        if decompiler.h_hint_args:
+            op_hstem = 'hstemhm' if mm_hint_info.hint_masks else 'hstem'
+            hint_program.extend(decompiler.h_hint_args)
+            hint_program.append(op_hstem)
+        if decompiler.v_hint_args:
+            hint_program.extend(decompiler.v_hint_args)
+            # Don't need to append op_vstem, as this is still in hint_program.
+            program = hint_program + program
+
+        # Re-calculate the hint masks.
+        if is_reference_font:
+            h_args = decompiler.h_hint_args
```

Comment:
The following piece of code for hhints is repeated for vhints and they look the same.
Maybe we may want to factor it out as a function?

Reply:
Good point. Done.